### PR TITLE
perf(expr): optimize casting to varchar

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR optimizes the performance of casting values to varchar.

It introduced write API for `ToText`, so that strings can be directly written to array buffers without generating String.
The display function of interval and timestampz was also optimized.

perf-cast

<details>
  <summary>Click to show full results</summary>

bench | Before time(us) | After time(us) | Change(%) | Speedup
-- | -- | -- | -- | --
cast(timestampz->varchar) | 508.640 | 121.600 | -76.1% | 3.2
cast(timestamp->varchar) | 166.200 | 58.245 | -65.0% | 1.9
cast(float64->varchar) | 78.386 | 57.597 | -26.5% | 0.4
cast(float32->varchar) | 57.903 | 37.384 | -35.4% | 0.5
cast(date->varchar) | 86.896 | 32.669 | -62.4% | 1.7
cast(time->varchar) | 47.508 | 28.428 | -40.2% | 0.7
cast(decimal->varchar) | 67.682 | 28.317 | -58.2% | 1.4
cast(int16->varchar) | 29.532 | 12.337 | -58.2% | 1.4
cast(int64->varchar) | 52.043 | 12.319 | -76.3% | 3.2
cast(int32->varchar) | 28.863 | 12.258 | -57.5% | 1.4
cast(boolean->varchar) | 26.826 | 6.396 | -76.2% | 3.2
bool_out(boolean) | 25.480 | 5.126 | -79.9% | 4.0

</details>

The `writer` argument of string functions was also changed from `StringWriter<'_>` to `&mut dyn Write`, making them decouple from array. I tried to use `&mut impl Write` but was blocked by annoying lifetime issues. Anyways, the performance of these operations is still slightly improved:

perf-string-ops

<details>
  <summary>Click to show full results</summary>

bench | Before time(us) | After time(us) | Change(%) | Speedup
-- | -- | -- | -- | --
rtrim(varchar,varchar) | 21.780 | 15.768 | -27.6% | 0.4
substr(varchar,int32,int32) | 11.126 | 8.090 | -27.3% | 0.4
rtrim(varchar) | 10.537 | 7.712 | -26.8% | 0.4
substr(varchar,int32) | 9.198 | 7.111 | -22.7% | 0.3
ltrim(varchar) | 9.661 | 8.010 | -17.1% | 0.2
trim(varchar) | 11.308 | 9.618 | -14.9% | 0.2
overlay(varchar,varchar,int32,int32) | 17.107 | 14.697 | -14.1% | 0.2
overlay(varchar,varchar,int32) | 13.408 | 12.007 | -10.4% | 0.1
ltrim(varchar,varchar) | 21.198 | 19.021 | -10.3% | 0.1
trim(varchar,varchar) | 20.876 | 19.205 | -8.0% | 0.1
split_part(varchar,varchar,int32) | 30.708 | 29.293 | -4.6% | 0.0
md5(varchar) | 346.010 | 331.670 | -4.1% | 0.0

</details>

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#6868

### CODE REVIEW

**Review by BowenXiao1999 — APPROVED** [review]

File: src/expr/src/expr/template.rs [hunk 1]
```diff
@@ -167,8 +164,9 @@ macro_rules! eval_bytes {
     ($self:ident, $output_array:ident, $($arg:ident,)*) => {
         if let ($(Some($arg), )*) = ($($arg, )*) {
             {
-                let writer = $output_array.writer();
-                let _guard = ($self.func)($($arg, )* writer)?;
+                let mut writer = $output_array.writer().begin();
+                let _guard = ($self.func)($($arg, )* &mut writer)?;
```
Comment (BugenZhao) [inline src/expr/src/expr/template.rs:168]:
```suggestion
                ($self.func)($($arg, )* &mut writer)?;
```

File: src/expr/src/vector_op/concat_op.rs [hunk 2]
```diff
@@ -12,31 +12,25 @@
...
 use crate::Result;
 
 #[inline(always)]
-pub fn concat_op(left: &str, right: &str, writer: StringWriter<'_>) -> Result<WrittenGuard> {
-    let mut writer = writer.begin();
-    writer.write_ref(left);
-    writer.write_ref(right);
-    Ok(writer.finish())
+pub fn concat_op(left: &str, right: &str, writer: &mut dyn Write) -> Result<()> {
+    writer.write_str(left).unwrap();
```
Comment (BugenZhao) [inline src/expr/src/vector_op/concat_op.rs:21]: How about removing the all old `write_ref` methods, including `BytesWriter`? 😄

File: src/expr/src/vector_op/cast.rs [hunk 3]
```diff
@@ -376,18 +377,23 @@ pub fn int32_to_bool(input: i32) -> Result<bool> {
 
 // For most of the types, cast them to varchar is similar to return their text format.
 // So we use this function to cast type to varchar.
-pub fn general_to_text<T: ToText>(elem: T) -> Result<Box<str>> {
-    Ok(elem.to_text().into_boxed_str())
+pub fn general_to_text(elem: impl ToText, writer: &mut dyn Write) -> Result<()> {
```
Comment (BugenZhao) [inline src/expr/src/vector_op/cast.rs:380]: I'm not sure whether `dyn` is good here. 🤔 Seems we're sure that this is the `PartialStringWriter` type.

**Review by BugenZhao — COMMENTED** [review]
Rest LGTM! 🚀

File: src/expr/src/vector_op/cast.rs (hunk 3 above)
Comment (wangrunji0408) [reply src/expr/src/vector_op/cast.rs:380]: Another type used is `String`. I tried `&mut impl Write` but failed. 😭

File: src/common/src/array/list_array.rs [hunk 4]
```diff
@@ -487,9 +487,10 @@ impl Debug for ListRef<'_> {
 impl ToText for ListRef<'_> {
     // This function will be invoked when pgwire prints a list value in string.
     // Refer to PostgreSQL `array_out` or `appendPGArray`.
-    fn to_text(&self) -> String {
+    fn write(&self, f: &mut dyn std::fmt::Write) -> std::fmt::Result {
```
[the remark this answers is not in the document]
Comment (wangrunji0408) [reply src/common/src/array/list_array.rs]: I changed it to generic. But the operation functions still have to be dynamic. Therefore I had to add a wrapper over `&mut dyn Write` in order to make it as`&mut impl Write`. 🤣

File: src/expr/src/vector_op/cast.rs (hunk 3 above)
Comment (BugenZhao) [reply src/expr/src/vector_op/cast.rs:381]: If there're two, we may use a type alias with `Either`? 🤣

Comment (BugenZhao) [reply src/expr/src/vector_op/cast.rs:381]: Then seems we don't need this dynamic wrapper anymore. 🤔

Comment (wangrunji0408) [reply src/expr/src/vector_op/cast.rs:381]: That looks really weird. I'd rather using dynamic. 🥵

Comment (BugenZhao) [reply src/expr/src/vector_op/cast.rs:381]: Actually, the performance of `eval_row` does not that matter. 🥵 Both LGTM.

File: src/expr/src/vector_op/cast.rs [hunk 5]
```diff
@@ -376,18 +378,32 @@ pub fn int32_to_bool(input: i32) -> Result<bool> {
...
+    elem.write(&mut DynWriter(writer)).unwrap();
```
Comment (BugenZhao) [inline src/expr/src/vector_op/cast.rs:382]:
```suggestion
pub fn general_to_text(elem: impl ToText, mut writer: &mut dyn Write) -> Result<()> {
    elem.write(&mut writer).unwrap();
```

Will this work?

**Review by BugenZhao — APPROVED** [review]

File: src/expr/src/vector_op/cast.rs (hunk 5 above)
Comment (wangrunji0408) [reply src/expr/src/vector_op/cast.rs]: It works! Amazing. 😵

**Pull request merged**